### PR TITLE
Cross platform

### PR DESCRIPTION
**Pull request by mola**
Add cmake files
* Change all for each to new c++11 for
* change __super

this plugins not ported:
- OrthoMap
- PathRoaming

Compile on Fedora 28 - osgEarth 2.9, osg 3.4

### CODE REVIEW

**Comment by tqjxlm** [comment]
I'm testing it on Ubuntu

**Comment by tqjxlm** [comment]
@mola 
Good job, it compiles on Ubuntu. The main program is able to run, too. But only basic plugins are loaded in my environment. I wonder if it is also the case on your machine.

It seems all the plugins that have extra dependencies (other than osg & osgEarth) fail to load. The dependency can be external libs or other plugins.

A log file can be found in /tmp/Atlas/Atlas/AtlasLog.txt
My errors write:

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libVRMode.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libVRMode.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libVRMode.so: undefined symbol: _ZTT18OpenVREventHandler)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureLine.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureLine.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureLine.so: undefined symbol: _ZTI8DrawLine)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureArea.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureArea.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libMeasureArea.so: undefined symbol: _ZTI11DrawPolygon)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libElevationSlice.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libElevationSlice.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libElevationSlice.so: undefined symbol: _ZTI15DrawSurfaceLine)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libCrowdSimulation.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libCrowdSimulation.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libCrowdSimulation.so: undefined symbol: _ZTI15DrawSurfaceLine)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libDrawSurfacePolygon.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libDrawSurfacePolygon.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libDrawSurfacePolygon.so: undefined symbol: _ZTI15DrawSurfaceLine)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libModelPath.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libModelPath.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libModelPath.so: undefined symbol: _ZTI15DrawSurfaceLine)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddOGCData.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddOGCData.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddOGCData.so: undefined symbol: _ZTI18EarthDataInterface)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddArcGISData.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddArcGISData.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddArcGISData.so: undefined symbol: _ZTI18EarthDataInterface)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddGDALData.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddGDALData.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddGDALData.so: undefined symbol: _ZTI18EarthDataInterface)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddXYZData.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddXYZData.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddXYZData.so: undefined symbol: _ZTI18EarthDataInterface)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddTMSData.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddTMSData.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libAddTMSData.so: undefined symbol: _ZTI18EarthDataInterface)"

[Qt]    [Warning]    Plugin loading failed: [ "/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libInsolationAnalysis.so" ]  "Cannot load library /home/tqjxlm/dev/Atlas/build-release/dist/plugins/libInsolationAnalysis.so: (/home/tqjxlm/dev/Atlas/build-release/dist/plugins/libInsolationAnalysis.so: undefined symbol: _ZTI10TileSelect)"

[Qt]    [Info]    Plugin dependency cannot be resolved:  "DiffAnalysisd"
[Qt]    [Info]    Plugin dependency cannot be resolved:  "MeasureTerrainAread"
[Qt]    [Info]    Plugin dependency cannot be resolved:  "MeasureTerrainVolumed"
[Qt]    [Info]    Plugin dependency cannot be resolved:  "ModelFlattend"

**Comment by mola** [comment]
Fix plugin dependencies

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+option(PRELOAD_MAP "Preload default map." OFF)
+
+if (PRELOAD_MAP)
+  add_definitions(-DPRELOAD_MAP)
+endif()
+
+set(DIST_PATH ${PROJECT_BINARY_DIR}/dist)
+set(DIST_PLUGINS_PATH ${DIST_PATH}/plugins)
+set(DIST_LIB_PATH ${DIST_PATH}/lib)
+set(DIST_RESOURCES_PATH ${DIST_PATH}/res)
```
Comment (tqjxlm) [inline CMakeLists.txt]: The folder renaming is not mentioned elsewhere. All resource path in the code refers to /resources. Maybe it's better to leave it unchanged?
`set(DIST_RESOURCES_PATH ${DIST_PATH}/resources)`

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  -losgSim
+  -losgParticle
+  -losgText
+  -losgUtil
+  -losgEarth
+  -losgEarthSymbology
+  -losgEarthFeatures
+  -losgEarthAnnotation
+  -losgEarthUtil
+  -lfribidi
```
Comment (tqjxlm) [inline CMakeLists.txt]: I'm not sure if this library is neccessary on Linux to properly show some special characters. If it is neccessary, I will make it optional later in another commit.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+foreach(p ${PLUGINS})
+  DefinePlugin(${p})
+endforeach()
+
+# Cleaning all previously built plugins.
+file(REMOVE_RECURSE ${DIST_PLUGINS_PATH})
+
+# Copying resources.
+file(GLOB resdata res/*)
```
Comment (tqjxlm) [inline CMakeLists.txt]: Here it also assumes the resource folder is under /res, however all the resources are under /resources dir.

**Review by tqjxlm — COMMENTED** [review]
A few minor changes need to be discussed.

**Review by tqjxlm — APPROVED** [review]
It compiles succefully. Most of the currently working plugins can function as expected. A few plugins broke because of Windows dependency. They will be rewritten in the future.

**Comment by tqjxlm** [comment]
[Linux support](https://trello.com/c/9kiXwUEl/13-linux-support)

**Comment by tqjxlm** [comment]
[CMake support](https://trello.com/c/pzet8jsJ/12-cmake-support)

**Comment by tqjxlm** [comment]
[Geodetic map support](https://trello.com/c/Y7vBh7AS/9-geodetic-map-support)

**Pull request merged**